### PR TITLE
fix(cache): web UI can read what the display service caches again

### PR DESCRIPTION
**Pull request by ChuckBuilds**
## Problem

On hdpi the web UI logged, over and over:

```
WARNING - Permission denied loading cache for display_current_state from /var/cache/ledmatrix/display_current_state.json: [Errno 13] Permission denied
```

The display service runs as root and the web interface as the installing user. Cache files are `0660`, so the web interface reads what the display writes only through the file's group — which the installers arrange with a `root:ledmatrix` directory carrying the setgid bit.

`ledmatrix-web.service` had `CacheDirectory=ledmatrix`. With `User=<installing user>`, systemd re-owns that directory **and everything in it** to the user and its primary group whenever the owner differs — for a directory root created, on the very first start. From then on every file root writes is `root:root 0660`.

On hdpi: directory `hdpi:hdpi` with no setgid, all 947 files re-owned at the same second of the 2026-09-14 boot, and 392 `root:root` files the web UI couldn't read — `display_current_state`, `display_on_demand_state`, and `plugin_metrics:*` for every plugin. ledpi, whose older installed unit lacks `CacheDirectory=`, is still `root:ledmatrix 2775` with no warnings.

Since #547, `install_service.sh` renders the web unit from the template, so **every fresh install** gets this. The startup validator's drift warning would also have pushed healthy rigs like ledpi into it.

## Fix

**Existing installs** only receive `git pull`, never a reinstalled unit, so their fix lives in the code the root display service runs:

- `DiskCache.set` gives each new file the directory's group (only when the directory is group-writable) and `0660`, on the open descriptor before `os.replace`. It no longer depends on setgid, and it closes a window where a fresh file was briefly visible as mkstemp's `0600`.
- `DiskCache.share_existing_files()` runs once per process from the cleanup thread and repairs files an older version left behind. It works through `O_NOFOLLOW` descriptors and skips hard links and other users' files: the directory is writable by the web user, so root must not be tricked into chown/chmod-ing something outside it.

**New installs:** the web unit drops `CacheDirectory=` / `CacheDirectoryMode=`, and `install_web_service.sh` stops replacing an existing directory's `ledmatrix` group with the user's own group (it now only fixes a `root`-group directory, and sets `2775`).

**Web status routes read the latest state** (second commit). This bug was hidden until the web UI could read these files, and testing on hdpi exposed it: `/display/on-demand/status` answered `"active"` for 100+ seconds while the file on disk said `"idle"`. Both `/display/on-demand/status` and `/display/current-status` read through the web process's memory tier, which serves its first copy for the full 120s `max_age`. They now pass `memory_ttl=0`, like every other cross-process reader (plugin health/metrics, the on-demand mailbox).

## Testing

- `test/test_cache_shared_group.py` (14 tests). All pass under WSL as root; as a regular user 12 pass and the 2 root-only tests skip. With the old code, 11–13 of them fail. Removing the `O_NOFOLLOW`/link-count guards makes the symlink and hard-link tests fail. The root-only end-to-end test forks a non-root reader, and it fails on `main`.
- **Real systemd in WSL**, with `setup_cache.sh` then `install_web_service.sh` as root with `SUDO_USER`:
  - `main`: the directory became `chuck:chuck`, root's file was `root:root`, and the user read failed.
  - This branch: the directory stayed `root:ledmatrix drwxrwsr-x` across restarts, root's file was `root:ledmatrix`, and the user read worked.
  - Migration from the broken state under the **old** unit: the repair made the files `root:chuck` and readable, and they stayed readable after a web restart.
  - Edge cases: no directory and user not in `ledmatrix` gave `root:<user> 2775`; an existing `root:root 755` directory gave `root:<user> 2775`.
- **Existing cache, systemd and installer test files:** the failure lists match `main` exactly on Windows (4) and WSL (23). Those failures come from test ordering and the harness.
- **hdpi**, with the two Python files dropped in and `ledmatrix.service` restarted:
  - The log shows `Made 392 cache file(s) ... readable by the directory's group`, and 0 `root:root` files remain.
  - `/api/v3/display/current-status` returns the live mode.
  - There have been 0 permission warnings since.

- **hdpi on this branch** (`git checkout` of the PR, old unit with `CacheDirectory=` still installed, both services restarted):
  - Overnight with the fix loaded: 0 permission warnings, 0 `root:root` files, and 27 new root-written files all `root:hdpi 660`.
  - A web restart under the old unit didn't re-own anything; `current-status`, `/plugins/metrics` (49 plugins) and `/plugins/health` all read the display's data.
  - Full on-demand cycle: about 16 seconds after start, `on-demand/status` = `active ledmatrix-stocks` and `current-status` = `stocks, on_demand_active: true`. Within 4 seconds of stop, both show `idle` and `odds_ticker`. The API matched the file on disk at every 4–5 second sample.
- `test/test_web_display_state_freshness.py`: two processes' caches over one directory. It fails without `memory_ttl=0`. `test_web_api.py` + `test_web_error_detail.py` all pass (82 before, 84 with the new tests).
- **Upgrading an existing rig** to the new unit is `sudo ./scripts/install/install_service.sh`, which the startup drift warning already suggests. It isn't required: the code fix works under the old unit.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved shared cache permissions so the display service and web interface can reliably access the same cache files.
  * Automatically repairs older cache files that were previously inaccessible to the web interface.
  * Fixed display status and on-demand status pages to reflect service changes immediately, without stale in-memory data.
* **Installation**
  * Updated service and cache directory setup to preserve shared access across installations and restarts.
* **Tests**
  * Added coverage for shared cache access, permission repair, and real-time display status updates.

### CODE REVIEW

File: scripts/install/install_web_service.sh [hunk 1]
```diff
@@ -62,31 +62,33 @@ for VERIFY_UNIT in ledmatrix-update-verify.service ledmatrix-update-verify.path;
...
+    chown root:"$CACHE_GROUP" "$CACHE_DIR" 2>/dev/null || true
     echo "✓ Cache directory created: $CACHE_DIR"
 else
-    # Ensure permissions are correct
-    chmod 775 "$CACHE_DIR" 2>/dev/null || true
-    # Try to set group ownership if possible
-    if getent group "$ACTUAL_USER" > /dev/null 2>&1; then
-        chown root:"$ACTUAL_USER" "$CACHE_DIR" 2>/dev/null || true
+    if [ "$(stat -c %g "$CACHE_DIR" 2>/dev/null)" = "0" ]; then
+        chgrp "$CACHE_GROUP" "$CACHE_DIR" 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (ChuckBuilds) [reply scripts/install/install_web_service.sh]: Partly adopted in the latest commit on this branch.

**Valid:** only a `root`-group directory was re-grouped. A directory in some other group the web user isn't in (e.g. `root:ledmatrix` for a user outside `ledmatrix`) was left alone, so root's files stayed unreadable. The script now changes the group whenever the installing user isn't a member of the directory's current group. When that happens and root-owned JSON files carrying the old group are present, it runs `systemctl try-restart ledmatrix.service`, so `DiskCache.share_existing_files()` re-groups them through its symlink- and hard-link-safe path rather than a recursive `chgrp`.

**Not adopted:** unconditionally `chown root:"$CACHE_GROUP"`. On a rig already working through the user's own group (the layout systemd's `CacheDirectory=` leaves, e.g. `hdpi:hdpi`), that would move the directory to `ledmatrix` and strand every existing file on the old group until the display restarts. A web process that is running but not yet in `ledmatrix` couldn't read root's new files at all until it restarted too. A directory whose group the user already belongs to is left as it is.

Verified under WSL systemd across 7 directory states: user group, ledmatrix member, ledmatrix non-member with and without root files, `root:root`, missing, and unnamed gid. The previous script left the non-member case unchanged.

[the remark this answers is not in the document]
Comment (ChuckBuilds) [reply scripts/install/install_web_service.sh]: Fixed in d5b5457f: the cache dir is re-grouped whenever the web user is not in its group, with a try-restart of ledmatrix.service so the safe repair re-groups root-owned JSON files.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
